syntax = "proto3";

package eddid;

import "typedef.proto";

// heartbeat
message Heartbeat
{
	int32 connection_id = 1;
}

message HeartbeatResp
{
	int32  error_code = 1;
    string error_msg = 2;
	string server_time = 3;
}

// 查询合约列表
message QueryAllInsInfo
{

}

message timeSpin
{
	int32 start = 1; // HHMMSS
	int32 end = 2;   // HHMMSS
}

message TradingTime
{
	repeated timeSpin time = 1;
}

message QuoteBasicInfo
{
	// 合约基本信息
	TypeDef.InstrumentType ins_type = 1; 	// 合约类型
	TypeDef.Exchange exchange = 2;			// 交易所ID
	string ins_number = 3;					// 合约序号
	string ins_id = 4;						// 合约ID
	string ins_name = 5;         			// 合约名称
	string ins_en_name = 6;        			// 合约英文名
	TypeDef.Currency settle_currency = 7;   // 结算货币
	TypeDef.Currency trade_currency = 8;	// 交易货币
	// 品种基本信息
	TypeDef.ProductType product_type = 9;   // 品种类型
    string product_code = 10;               // 品种代码
	string product_name = 11;               // 品种名称
	string product_en_name = 12;			// 品种英文名称
	TradingTime trading_time = 13;          // 交易时间

	int32 lot_size = 14;                  	// 每手数量
	double min_change_price_gap = 15;       // 最小变动价差
	bool is_trade = 16;                     // 是否可交易
	TypeDef.InstrumentStatus ins_status = 17; // 合约状态
	string create_date = 18;                  // 创建日期
	string listing_date = 19;				  // 上市日期 YYYYMMDD
	string update_date = 20;                  // 更新日期
	// 期货合约私有信息
	TypeDef.RateType deposit_collect_method = 21; // 保证金收取方式
	double long_deposit_rate = 22;            // 多头保证金率
	double short_deposit_rate = 23;           // 空头保证金率
	int32 market_price_max_vol = 24;          // 市价单最大数量
	int32 market_price_min_vol = 25;          // 市价单最小数量
	int32 limit_price_max_vol = 26;           // 限价单最大数量
	int32 limit_price_min_vol = 27;           // 限价单最小数量
	double ins_multiplier = 28;               // 合约乘数
	int32 delivery_year = 29;                 // 交割年
	int32 delivery_mouth = 30;                // 交割月
	bool is_expired = 31;                     // 是否过期
	bool is_main_ins = 32;                    // 是否是主力合约
	string trade_last_date = 33;              // 最后交易日
	string first_notice_date = 34;            // 首次通知日
	string expired_date = 35;                 // 到期日
	int32 expired_mouth = 36;                 // 到期月份
	string start_delivery_date = 37;          // 开始交割日期 YYYYMMDD
	string end_delivery_date = 38;            // 结束交割日期 YYYYMMDD
};

message QueryAllInsInfoResp
{
	int32  error_code = 1;
    string error_msg = 2;
	repeated QuoteBasicInfo data = 3;
}

// 订阅行情

message CodeInfo
{
	TypeDef.SubType type = 1;       		// 订阅类型
	TypeDef.Exchange exchange = 2;			// 交易所ID
	string ins_id = 3;						// 合约ID
}

message SubQuotes
{
	repeated CodeInfo info = 1;
}

message SubQuotesResp
{
	int32  error_code = 1;
    string error_msg = 2;
}

// 取消订阅行情
message UnSubQuotes
{
	repeated CodeInfo info = 1;
}

message UnSubQuotesResp
{
	int32  error_code = 1;
    string error_msg = 2;
}

// 查询K线
message QueryKLine
{
	TypeDef.KLineType type = 1;				// K线类型
	TypeDef.Exchange exchange = 2;			// 交易所ID
	string ins_id = 3;						// 合约ID
	string start_time = 4;                  // 开始时间
	string end_time = 5;                    // 结束时间
}

message KLineInfo
{
	string time = 1;   				// 时间
	double open = 2;   				// 开盘价
	double high = 3;   				// 最高价
	double low = 4;    				// 最低价
	double close = 5;  				// 收盘价
	double total_amount = 6;       // 成交总金额
    int64 total_volume = 7;        // 成交总量
}

message QueryKLineResp
{
	int32  error_code = 1;
    string error_msg = 2;
    TypeDef.Exchange exchange = 3; // 交易所ID
    string product_code = 4;       // 品种代码
    string ins_id = 5;             // 合约ID
	TypeDef.KLineType type = 6;	   // K线类型
	repeated KLineInfo data = 7;   // kline数据
}

// 查询经纪人队列
message QueryBroker
{
	TypeDef.Exchange exchange = 1;			// 交易所ID
	string ins_id = 2;						// 合约ID
	TypeDef.Language language = 3;          // 返回证券名字的语言
}

message BrokerInfo
{
	string code = 1; // 证券代码
	string name = 2; // 证券名字
}

message Brokers
{
	int32 level = 1;
	repeated BrokerInfo info = 2;
}

message QueryBrokerResp
{
	int32  error_code = 1;
    string error_msg = 2;
    TypeDef.Exchange exchange = 3; // 交易所ID
    string product_code = 4;       // 品种代码
    string ins_id = 5;             // 合约ID
	repeated Brokers buy = 7;      // 买盘经纪人
	repeated Brokers sell = 8;     // 卖盘经纪人 
}

// 查询最近交易日
message QueryTradeDate
{

}

message TradeDateInfo
{
	TypeDef.Exchange exchange = 1;	// 交易所ID
	repeated int32 date = 2;  		// 交易日数组 YYYYMMDD
}

message QueryTradeDateResp
{
	int32  error_code = 1;
    string error_msg = 2;
	repeated TradeDateInfo data = 3;
}

